subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheCompositeBuildsIntegrationTest.groovy

```diff
@@ -270,6 +271,60 @@ class ConfigurationCacheCompositeBuildsIntegrationTest extends AbstractConfigura
         }
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/20945")
+    def "composite build with dependency substitution can include builds in any order"() {
+        given:
+        order.each {
+            settingsFile "includeBuild '$it'\n"
+        }
+        buildFile '''
+            ['clean', 'build'].each { name ->
+                tasks.register(name) {
+                    gradle.includedBuilds.each { build ->
+                        dependsOn(build.task(':' + name))
+                    }
+                }
+            }
+        '''
+        createDir('lib') {
+            file('settings.gradle') << 'rootProject.name = "lib"'
+            file('build.gradle') << '''
+                plugins { id 'java-library' }
+                group = 'com.example'
+                version = '1.0'
+            '''
+        }
+        createDir('util') {
+            file('settings.gradle') << 'rootProject.name = "util"'
+            file('build.gradle') << '''
+                plugins { id 'java-library' }
+                dependencies {
+                    api 'com.example:lib:1.0'
+                }
+            '''
+        }
+
+        def configurationCache = newConfigurationCacheFixture()
+
+        when:
+        configurationCacheRun 'clean', 'build'
+
+        then:
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheRun 'clean', 'build'
+
+        then:
+        configurationCache.assertStateLoaded()
+
+        and:
+        result.assertTaskOrder(':lib:build', ':util:build', ':build')
```

Comment:
Looks like having a dependency [isn't enough](https://ge.gradle.org/scans/tests?search.timeZoneId=Europe/Berlin&tests.container=org.gradle.configurationcache.ConfigurationCacheCompositeBuildsIntegrationTest&tests.test=composite%20build%20with%20dependency%20substitution%20can%20include%20builds%20in%20any%20order%20%5Border:%20%5Blib%2C%20util%5D%2C%20%230%5D) to ensure the declared order of `:build` tasks. I suppose, the exported configuration of `lib` is built by `assemble` or some other lower-level task.

Reply:
Indeed. `compileJava` is the right one.